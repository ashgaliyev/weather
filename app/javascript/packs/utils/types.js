import PropTypes from "prop-types";

export const weatherShape = PropTypes.shape({
  tempMin: PropTypes.number.isRequired,
  tempMax: PropTypes.number.isRequired,
  icon: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  windSpeed: PropTypes.number.isRequired,
  date: PropTypes.string.isRequired,
});

export const fiveDaysForecastShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  fiveDays: PropTypes.arrayOf(weatherShape),
});

export const currentForecastShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  weather: weatherShape,
});

export const placeShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  forecastId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
});
