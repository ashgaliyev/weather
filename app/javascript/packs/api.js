import axios from "axios";

export const createPlace = (place) =>
  new Promise((resolve, reject) => {
    axios({
      method: "POST",
      url: "/places",
      data: { place: place },
      headers: {
        "X-CSRF-Token": document.querySelector("meta[name=csrf-token]").content,
      },
    })
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        if (error.response.status === 422) {
          resolve(error.response);
        } else {
          reject(error);
        }
      });
  });

export const updatePlace = (place) =>
  new Promise((resolve, reject) => {
    const { id, ...placeData } = place;
    axios({
      method: "PUT",
      url: `/places/${id}`,
      data: { place: placeData },
      headers: {
        "X-CSRF-Token": document.querySelector("meta[name=csrf-token]").content,
      },
    })
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        if (error.response.status === 422) {
          resolve(error.response);
        } else {
          reject(error);
        }
      });
  });

export const deletePlace = (place) =>
  new Promise((resolve, reject) => {
    axios({
      method: "DELETE",
      url: `/places/${place.id}`,
      headers: {
        "X-CSRF-Token": document.querySelector("meta[name=csrf-token]").content,
      },
    })
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });

export const updateTempUnit = (unit) =>
  new Promise((resolve, reject) => {
    axios({
      method: "PUT",
      url: "/update_temp_unit",
      data: { temp_unit: unit },
      headers: {
        "X-CSRF-Token": document.querySelector("meta[name=csrf-token]").content,
      },
    })
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
