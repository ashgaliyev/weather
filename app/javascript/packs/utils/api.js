import axios from "axios";

const csrfToken = () => {
  const element = document.querySelector("meta[name=csrf-token]");
  if (element) {
    return element.content;
  }
  return null;
};

export const createPlace = place =>
  new Promise((resolve, reject) => {
    axios({
      method: "POST",
      url: "/places",
      data: { place },
      headers: {
        "X-CSRF-Token": csrfToken()
      }
    })
      .then(response => {
        resolve(response);
      })
      .catch(error => {
        if (error.response.status === 422) {
          resolve(error.response);
        } else {
          reject(error);
        }
      });
  });

export const updatePlace = place =>
  new Promise((resolve, reject) => {
    const { id, ...placeData } = place;
    axios({
      method: "PUT",
      url: `/places/${id}`,
      data: { place: placeData },
      headers: {
        "X-CSRF-Token": csrfToken()
      }
    })
      .then(response => {
        resolve(response);
      })
      .catch(error => {
        if (error.response.status === 422) {
          resolve(error.response);
        } else {
          reject(error);
        }
      });
  });

export const deletePlace = place =>
  new Promise((resolve, reject) => {
    axios({
      method: "DELETE",
      url: `/places/${place.id}`,
      headers: {
        "X-CSRF-Token": csrfToken()
      }
    })
      .then(response => {
        resolve(response);
      })
      .catch(error => {
        reject(error);
      });
  });

export const updateTempUnit = unit =>
  new Promise((resolve, reject) => {
    axios({
      method: "PUT",
      url: "/update_temp_unit",
      data: { temp_unit: unit },
      headers: {
        "X-CSRF-Token": csrfToken()
      }
    })
      .then(response => {
        resolve(response);
      })
      .catch(error => {
        reject(error);
      });
  });
