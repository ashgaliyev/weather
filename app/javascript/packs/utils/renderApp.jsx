import { createRoot } from "react-dom/client";
import * as React from "react";
import SettingsProvider from "../components/SettingsProvider";

const renderApp = App => {
  const props = window.props ? JSON.parse(window.props) : {};
  const settings = window.settings ? JSON.parse(window.settings) : {};

  const container = document.getElementsByTagName("main")[0];
  const root = createRoot(container);

  root.render(
    <SettingsProvider {...settings}>
      <App {...props} />
    </SettingsProvider>
  );
};

export default renderApp;
