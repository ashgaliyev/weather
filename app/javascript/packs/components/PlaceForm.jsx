import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import Map from "./Map";
import { createPlace, updatePlace } from "../utils/api";
import { placeFormShape } from "../utils/types";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  place: placeFormShape,
};

const LABEL = {
  name: "Name",
  lat: "Latitude",
  lng: "Longitude",
};

const PlaceForm = ({ onSubmit, place }) => {
  const placeIsSet = place !== undefined;

  const [form, setForm] = useState({
    name: placeIsSet ? place.name : "",
    lat: placeIsSet ? place.lat : "",
    lng: placeIsSet ? place.lng : "",
  });

  const [invalidFields, setInvalidFields] = useState({
    name: null,
    lat: null,
    lng: null,
  });

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setForm((prevForm) => ({
      ...prevForm,
      [name]: value,
    }));
  };

  const handleGeoCode = ({ lat, lng, address }) => {
    setForm({
      lat,
      lng,
      name: address,
    });
  };

  const handleFormSubmit = () => {
    const { name, lat, lng } = form;
    if (placeIsSet) {
      updatePlace({
        id: place.id,
        name,
        lat,
        lng,
      }).then((response) => {
        if (response.status === 422) {
          setInvalidFields(response.data);
        } else {
          onSubmit(response.place);
        }
      });
    } else {
      createPlace({
        name,
        lat,
        lng,
      }).then((response) => {
        if (response.status === 422) {
          setInvalidFields(response.data);
        } else {
          onSubmit(response.place);
        }
      });
    }
  };

  return (
    <div className="w-full flex flex-col">
      <form
        className="flex flex-row mb-5"
        onSubmit={(e) => {
          e.preventDefault();
          setInvalidFields({
            name: null,
            lat: null,
            lng: null,
          });
          handleFormSubmit();
        }}
      >
        {["name", "lat", "lng"].map((key, i) => {
          const value = form[key];
          return (
            <div key={i} className="flex flex-col mr-3">
              <label htmlFor={key}>{LABEL[key]}</label>
              <input
                type="text"
                name={key}
                id={key}
                value={value}
                onChange={handleInputChange}
              />
              {invalidFields[key] && (
                <span className="text-red-500 text-sm">
                  {invalidFields[key].join(", ")}
                </span>
              )}
            </div>
          );
        })}

        <div className="flex flex-col justify-start mt-6">
          <button
            className="bg-blue-500 text-white p-2 rounded"
            type="submit"
            data-test-id="submit-place"
          >
            {placeIsSet ? "Update" : "Create"}
          </button>
        </div>
      </form>
      <Map
        onGeocode={handleGeoCode}
        defaultCenter={
          placeIsSet
            ? {
                lat: place.lat,
                lng: place.lng,
              }
            : null
        }
      />
    </div>
  );
};

PlaceForm.propTypes = propTypes;

export default PlaceForm;
