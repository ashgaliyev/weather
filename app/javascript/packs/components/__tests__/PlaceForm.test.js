import React from "react";
import renderer from "react-test-renderer";
import { cleanup, fireEvent, render, act } from "@testing-library/react";
import PlaceForm from "../PlaceForm";
import { createPlace } from "../../utils/api";

jest.mock("../../utils/api", () => ({
  createPlace: jest.fn(),
  updatePlace: jest.fn().mockResolvedValue(() =>
    Promise.resolve({
      response: { status: 200 },
    })
  ),
}));

describe("PlaceForm", () => {
  beforeEach(() => {
    React.useContext = jest.fn().mockReturnValue({ mapApiKey: null });
  });

  describe("when place provided", () => {
    it("renders correctly", () => {
      const props = {
        place: {
          id: 1,
          name: "Paris",
          lat: 48.8566,
          lng: 2.3522,
        },
        onSubmit: jest.fn(),
      };

      const tree = renderer.create(<PlaceForm {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("when place not provided", () => {
    it("renders correctly", () => {
      const props = {
        onSubmit: jest.fn(),
      };

      const tree = renderer.create(<PlaceForm {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("form submission", () => {
    it("enters place data and submits", async () => {
      const onSubmit = jest.fn();

      const place = {
        name: "Paris",
        lat: "48.8566",
        lng: "2.3522",
      };

      createPlace.mockResolvedValue(
        Promise.resolve({
          status: 200,
          place,
        })
      );

      const { getByLabelText, getByText } = render(
        <PlaceForm onSubmit={onSubmit} />
      );
      const nameInput = getByLabelText("Name");
      const latInput = getByLabelText("Latitude");
      const lngInput = getByLabelText("Longitude");

      fireEvent.change(nameInput, { target: { value: place.name } });
      fireEvent.change(latInput, { target: { value: place.lat } });
      fireEvent.change(lngInput, { target: { value: place.lng } });

      await fireEvent.click(getByText("Create"));

      expect(onSubmit).toHaveBeenCalledWith(place);
    });

    it("enters place data and submits with errors", async () => {
      const onSubmit = jest.fn();

      const place = {
        name: "Paris",
        lat: "48.8566",
        lng: "2.3522",
      };

      createPlace.mockResolvedValue(
        Promise.resolve({
          status: 422,
          data: { name: ["can't be blank"] },
        })
      );

      const { getByLabelText, getByText } = render(
        <PlaceForm onSubmit={onSubmit} />
      );
      const nameInput = getByLabelText("Name");
      const latInput = getByLabelText("Latitude");
      const lngInput = getByLabelText("Longitude");

      fireEvent.change(nameInput, { target: { value: place.name } });
      fireEvent.change(latInput, { target: { value: place.lat } });
      fireEvent.change(lngInput, { target: { value: place.lng } });

      await act(async () => {
        await fireEvent.click(getByText("Create"));
      });

      expect(getByText("can't be blank")).toBeTruthy();
      expect(onSubmit).not.toHaveBeenCalled();
    });
  });
});
