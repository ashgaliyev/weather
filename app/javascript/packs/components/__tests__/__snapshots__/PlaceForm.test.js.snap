// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceForm when place not provided renders correctly 1`] = `
<div
  className="w-full flex flex-col"
>
  <form
    className="flex flex-row mb-5"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-col mr-3"
    >
      <label
        htmlFor="name"
      >
        Name
      </label>
      <input
        id="name"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="flex flex-col mr-3"
    >
      <label
        htmlFor="lat"
      >
        Latitude
      </label>
      <input
        id="lat"
        name="lat"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="flex flex-col mr-3"
    >
      <label
        htmlFor="lng"
      >
        Longitude
      </label>
      <input
        id="lng"
        name="lng"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="flex flex-col justify-start mt-6"
    >
      <button
        className="bg-blue-500 text-white p-2 rounded"
        data-test-id="submit-place"
        type="submit"
      >
        Create
      </button>
    </div>
  </form>
  <span>
    No API key provided
  </span>
</div>
`;

exports[`PlaceForm when place provided renders correctly 1`] = `
<div
  className="w-full flex flex-col"
>
  <form
    className="flex flex-row mb-5"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-col mr-3"
    >
      <label
        htmlFor="name"
      >
        Name
      </label>
      <input
        id="name"
        name="name"
        onChange={[Function]}
        type="text"
        value="Paris"
      />
    </div>
    <div
      className="flex flex-col mr-3"
    >
      <label
        htmlFor="lat"
      >
        Latitude
      </label>
      <input
        id="lat"
        name="lat"
        onChange={[Function]}
        type="text"
        value={48.8566}
      />
    </div>
    <div
      className="flex flex-col mr-3"
    >
      <label
        htmlFor="lng"
      >
        Longitude
      </label>
      <input
        id="lng"
        name="lng"
        onChange={[Function]}
        type="text"
        value={2.3522}
      />
    </div>
    <div
      className="flex flex-col justify-start mt-6"
    >
      <button
        className="bg-blue-500 text-white p-2 rounded"
        data-test-id="submit-place"
        type="submit"
      >
        Update
      </button>
    </div>
  </form>
  <span>
    No API key provided
  </span>
</div>
`;
